tablo/interfaces/arcgis/views.py

```diff
@@ -259,44 +266,61 @@ def handle_request(self, request, **kwargs):
         except (DatabaseError, ValueError):
             return HttpResponseBadRequest(json.dumps({'error': 'Invalid request'}))
 
+        # Process limited query before calculating totals and counts
+        exceeded_limit = len(query_response) > int(limit)
+        query_response = query_response[:-1] if exceeded_limit else query_response
+
         if return_count_only:
-            response = query_response[0]
+            data = query_response[0]
+        elif return_format == 'csv':
+            header = list(query_response[0].keys())
+            header.sort()
+
+            data = [[h.strip('"').join('""') for h in header]] if offset == 0 else []
+            for item in query_response:
+                data.append([str(item[field]).strip('"').join('""') for field in header])
         else:
-            record_count = len(query_response)
-            exceeded_limit = record_count > int(limit)
-            query_response = query_response[:-1] if exceeded_limit else query_response
-
-            response = {
+            data = {
                 'count': len(query_response),
                 'total': len(query_response),
                 'exceededTransferLimit': exceeded_limit
             }
             if return_ids_only:
                 object_id_field = self.feature_service_layer.object_id_field
-                response['objectIdFieldName'] = object_id_field
-                response['objectIds'] = [feature[object_id_field] for feature in query_response]
+                data['objectIdFieldName'] = object_id_field
+                data['objectIds'] = [feature[object_id_field] for feature in query_response]
             else:
                 features = convert_wkt_to_esri_feature(query_response, self.feature_service_layer)
-                response.update({
+                data.update({
                     'count': len(features),        # Root level (source) features
                     'total': len(query_response),  # Total number of records queried
                     'fields': self.feature_service_layer.fields,
                     'geometryType': 'esriGeometryPoint',
                     'features': features
                 })
 
-                if response['total'] > response['count']:
-                    response['relatedFields'] = {
+                if data['total'] > data['count']:
```

Comment:
I created a new JIRA issue, DB-5191, to (1) rework when related fields may be queried, (2) to revisit the limit/offset implementation, and (3) to remove total/count.

I made it a new issue because there have been significant new changes to this branch that stretch the scope, and because this branch was created from DB-4919, changes to that branch will cause difficult merge conflicts. Also, @MikeTheReader has been using this branch in testing, and for the time being this piece isn't causing a problem. The new issue is in version 3.1, and I queued it toward the top of the DWR epic.